src/misc/firespread.ts

```diff
@@ -0,0 +1,71 @@
+import { Material } from 'org.bukkit';
+import { Block } from 'org.bukkit.block';
+import { Fire } from 'org.bukkit.block.data.type';
+import { BlockBurnEvent, BlockSpreadEvent } from 'org.bukkit.event.block';
+
+const MAX_FIRE_AGE = 15; // CANNOT BE MORE THAN 15 IN VANILLA -> Errors
+const VERTICAL_MULTIPLIER = 0.98;
+const HORIZONTAL_MULTIPLIER = 0.3;
+const OLD_EXTINGUISH_CHANCE = 0.75; // Fires might not extinguish completely if below 1-HORIZONTAL_MULTIPLIER
+
+const FIREPLACE_BLOCKS = [Material.COAL_BLOCK, Material.MAGMA_BLOCK];
+
+const chanceOf = (percent: number) => Math.random() < percent;
+
+function addAge(fire: Block) {
+  if (fire.type !== Material.FIRE) return;
+  const fireData = fire.blockData as Fire;
+  if (fireData.age + 1 >= MAX_FIRE_AGE) {
+    fireData.age = MAX_FIRE_AGE;
+  } else {
+    fireData.age++;
+  }
+  fire.blockData = fireData;
+}
+
+function getAge(fire: Block): number {
+  const fireData = fire.blockData as Fire;
+  return fireData.age;
+}
+
+function cancelFireSpreadByChance(source: Block, dirMod: number): boolean {
+  if (getAge(source) <= 1) {
```

Comment:
Is it intentional that fire spread slows down between ages 1 and 5? If it is, please describe that in a comment.